### PR TITLE
center map on substation on user clicks

### PR DESCRIPTION
**Pull request by jonenst**
Signed-off-by: Jon Harper <jon.harper87@gmail.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
When clicking on a voltage level, only the svg is update

**What is the new behavior (if this is a feature change)?**
the map is centered on the substation of the voltage level

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
NO

**Other information**:
We can choose what to do when we close the svg, either nothing or zoom back to the whole network. For simplicity, I zoomed back to the whole network.
* We can also choose to not center on a substation during page load if we want to.
* We can also choose to no center on a substation when clicking on the map.
* We can also choose to not center on a substation on svg clicks.
* We can also choose to center on substation not when clicking on a vl in the list, but on a small button next to it, displayed only when selected (or hovered) ?

### CODE REVIEW

**Comment by geofjamg** [comment]
I would not do anything when closing the voltage level diagram, just let the center on last opened voltage level.

**Comment by geofjamg** [comment]
Also would it be feasible to flyby to the voltage level instead of going directly to it?

**Comment by jonenst** [comment]
> I would not do anything when closing the voltage level diagram, just let the center on last opened voltage level.

yeah you're right, done.

> Also would it be feasible to flyby to the voltage level instead of going directly to it?

Yes, I tried it and it just worked™.  You just have to add "transitionDuration: 1000" to the viewstate.


What do we want on page refresh with a voltagelevelid in the url ? Center on the whole network instantly like when there is no voltage level ? Center on the substation instantly ? Center on the substation with fly-by (this is what the code currently does)?

**Comment by geofjamg** [comment]
> > I would not do anything when closing the voltage level diagram, just let the center on last opened voltage level.
> 
> yeah you're right, done.
> 
> > Also would it be feasible to flyby to the voltage level instead of going directly to it?
> 
> Yes, I tried it and it just worked™. You just have to add "transitionDuration: 1000" to the viewstate.
> 
> What do we want on page refresh with a voltagelevelid in the url ? Center on the whole network instantly like when there is no voltage level ? Center on the substation instantly ? Center on the substation with fly-by (this is what the code currently does)?

I would center on the substation without flyby.

**Comment by jonenst** [comment]
> > What do we want on page refresh with a voltagelevelid in the url ? 

> I would center on the substation without flyby.

Done

**Comment by jonenst** [comment]
I also added the FlyTo interpolator, how do you like it ? When you are zoomed in, it flies out to about the starting zoom level before flying back to the previous zoom level on the new substation

**Comment by jonenst** [comment]
I changed the duration to 2s, it seems to be a good middle ground value. A bit slow when zoomed out, but not too fast when zoomed in

**Comment by jonenst** [comment]
If that's ok, we can merge @geofjamg ?

**Comment by geofjamg** [comment]
> If that's ok, we can merge @geofjamg ?

I will test again but I am not sure it was a good idea and a good feature.
I recommand you to test with a large network, it is slow and it also slow down diagram opening I suspect.

**Comment by geofjamg** [comment]
> > If that's ok, we can merge @geofjamg ?
> 
> I will test again but I am not sure it was a good idea and a good feature.
> I recommand you to test with a large network, it is slow and it also slow down diagram opening I suspect.

Try to quickly click on various voltage levels and you will see that diagram opening is totally blocked by network map flyby.

**Comment by geofjamg** [comment]
Maybe we should not track substation move automatically but on demand

**Comment by geofjamg** [comment]
Or maybe we could let it as it is and add a parameter to activate it or not (not by default) and see if people like it even if it slow down substation browsing.

**Comment by geofjamg** [comment]
Why did you commit this on jon PR?

**Comment by mikatheo** [comment]
> Why did you commit this on jon PR?

Unless I'm mistaken, we are on the Jon PR.

**Comment by geofjamg** [comment]
Also I would make the target appears only on list item hover, that is a common to propose additional functionalities without cluttering the UI with lot of useless buttons.

**Comment by mikatheo** [comment]
> Also I would make the target appears only on list item hover, that is a common to propose additional functionalities without cluttering the UI with lot of useless buttons.

Okay, but we have to keep in mind that we will have to make this responsive later. This feature won't be usable on Mobile at the moment.

**Comment by geofjamg** [comment]
> > Also I would make the target appears only on list item hover, that is a common to propose additional functionalities without cluttering the UI with lot of useless buttons.
> 
> Okay, but we have to keep in mind that we will have to make this responsive later. This feature won't be usable on Mobile at the moment.

hmmm right I didn't this in mind anymore about responsiveness. What is the right way to do it on mobile?

**Comment by mikatheo** [comment]
> > > Also I would make the target appears only on list item hover, that is a common to propose additional functionalities without cluttering the UI with lot of useless buttons.
> > 
> > 
> > Okay, but we have to keep in mind that we will have to make this responsive later. This feature won't be usable on Mobile at the moment.
> 
> hmmm right I didn't this in mind anymore about responsiveness. What is the right way to do it on mobile?

On closer inspection, an action (or a secondary actions button) for all items of a list seems to be Material-design compliant. This is the only way it will always be available whatever the device used.  
image
Perhaps it will seem more pleasing to the eye if the density of substation occurrences is reduced? There could also be later subheading additional information such as country and voltage level?

File: src/components/study-pane.js [hunk 1]
```diff
@@ -79,7 +79,9 @@ const StudyPane = () => {
 
     const [studyNotFound, setStudyNotFound] = useState(false);
 
-    const [voltageLevelId, setVoltageLevelId] = useState(null);
+    const [displayedVoltageLevelId, setDisplayedVoltageLevelId] = useState(null);
+
+    const [focusedVoltageLevelId, setFocusedVoltageLevelId] = useState(null);
```
Comment (geofjamg) [inline src/components/study-pane.js:84]: What is the purpose of this variable?

Comment (mikatheo) [reply src/components/study-pane.js:84]: "voltageLeveld" was initially used to open the diagram from the clicked voltage level in the explorer. Now I have also to share an another ID for the map from an action in the explorer. That's why there are now two variables in the study-pane common context. Do you have a suggestion to share this information ?

**Comment by mikatheo** [comment]
Note for myself : internationalize the 2 labels

File: src/components/study-pane.js (hunk 1 above)
Comment (geofjamg) [reply src/components/study-pane.js:84]: ok I understand

File: src/components/study-pane.js [hunk 2]
```diff
@@ -184,15 +186,21 @@ const StudyPane = () => {
...
     } else {
-        let voltageLevel = null;
-        if (voltageLevelId && network) {
-            voltageLevel = network.getVoltageLevel(voltageLevelId);
+        let displayedVoltageLevel, focusedVoltageLevel = null;
+        if (network) {
+            if (displayedVoltageLevelId) {
+                displayedVoltageLevel = network.getVoltageLevel(displayedVoltageLevelId);
+            }
+            if (focusedVoltageLevelId) {
```
Comment (geofjamg) [inline src/components/study-pane.js:194]: You don't need this variable in the renderer?

File: src/components/network/network-map.js [hunk 3]
```diff
@@ -62,39 +64,64 @@ const NetworkMap = (props) => {
     // we get the ref to the deck and it has not yet initialized..)
     function onAfterRender() {
             //use centered and deck to execute this block only once when the data is ready and deckgl is initialized
-            if (!centered && deck !== null && deck.viewManager != null && props.geoData !== null) {
+            //TODO, replace the next lines with setProps( { initialViewState } ) when we upgrade to 8.1.0
+            //see https://github.com/uber/deck.gl/pull/4038
+            //This is a hack because it accesses the properties of deck directly but for now it works
+            if ((!centered.centered || (props.centeredSubstationId && props.centeredSubstationId != centered.lastCenteredSubstation))
```
Comment (geofjamg) [inline src/components/network/network-map.js:70]: This prop should be added to propTypes and defaultProps

File: src/components/network/network-explorer.js [hunk 4]
```diff
@@ -59,22 +71,61 @@ const NetworkExplorer = ({network, onVoltageLevelClick}) => {
...
+        if (onVoltageLevelFocusClick !== null) {
+            const vl = filteredVoltageLevels[voltageLevelMenuIndex];
+            onVoltageLevelFocusClick(vl.id);
+        }
+        handleClose();
+    }
+
+    const voltagelevelInfo = index => {
+        if (network.getSubstation(filteredVoltageLevels[index].substationId)) {
+            return network.getSubstation(filteredVoltageLevels[index].substationId).countryName + " - " + filteredVoltageLevels[index].nominalVoltage + " kV";
```
Comment (jonenst) [inline src/components/network/network-explorer.js:106]: maybe put the voltage first as it is always short. If we ellipsize it will still be visible

**Pull request merged**